File: src/main/java/com/example/demo/service/component/LockManager.java
package com.example.demo.service.component;

import java.util.concurrent.TimeUnit;
import java.util.function.Supplier;

import com.example.demo.service.exceptions.LockAcquisitionException;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.redisson.api.RLock;
import org.redisson.api.RedissonClient;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

@Component("LockManager")
public class LockManager {

    static final Logger LOGGER = LogManager.getLogger(LockManager.class.getName());
    private final RedissonClient client;
    private final Long lockWaitTime;
    private final Long autoUnlockTime;

    public LockManager(RedissonClient client,
                       @Value("${redis.lock.wait.time.seconds}") Long lockWaitTime,
                       @Value("${redis.auto.unlock.time.seconds}") Long autoUnlockTime) {
        this.client = client;
        this.lockWaitTime = lockWaitTime;
        this.autoUnlockTime = autoUnlockTime;
    }

    public <T> T applyLock(String key, Supplier<T> supplier) {

        LOGGER.info("Locking {}...", key);
        RLock balance = client.getLock(key);
        try {
            boolean locked = balance.tryLock(lockWaitTime, autoUnlockTime, TimeUnit.SECONDS);
            if (!locked) {
                String msg = String.format("Lock can not be getting after %s seconds", lockWaitTime);
                throw new LockAcquisitionException(msg);
            }
            LOGGER.info("{} locked.", key);
            return supplier.get();
        } catch (InterruptedException e) {
            String msg = String.format("Error locking with key [%s]", key);
            LOGGER.error(msg, e);

            Thread.currentThread().interrupt();

            throw new RuntimeException(msg, e);
        } finally {
            if (balance.isLocked() && balance.isHeldByCurrentThread()) {
                LOGGER.info("Unlocking {}...", key);
                balance.unlock();
            }
            LOGGER.info("{} unlocked.", key);
        }
    }
}

File: src/main/kotlin/com/pepper/racecondition/redisson/lock/kotlin/LockManager.kt
package com.pepper.racecondition.redisson.lock.kotlin

import com.pepper.racecondition.redisson.lock.LockAcquisitionException
import com.pepper.racecondition.redisson.lock.LockInterruptException
import org.apache.logging.log4j.LogManager
import org.redisson.api.RedissonClient
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit

@Component("LockManagerKotlin")
class LockManager(private val client: RedissonClient,
                  @Value("\${redis.lock.wait.time.seconds}") private val lockWaitTime: Long,
                  @Value("\${redis.auto.unlock.time.seconds}") private val autoUnlockTime: Long) {

    private val logger = LogManager.getLogger(this::class.java)

    @Throws(LockInterruptException::class)
    fun <T> applyLock(key: String, supplier: () -> (T)): T {

        logger.info("Locking {}...", key)
        val lockInstance = client.getLock(key)

        try {
            val locked = lockInstance.tryLock(lockWaitTime, autoUnlockTime, TimeUnit.SECONDS)
            if (!locked) {
                logger.info("{} locked.", key)
                val msg = String.format("Lock can not be getting after %s seconds", lockWaitTime)
                throw LockAcquisitionException(msg)
            }else{
                logger.info("run *****************************", key)
                return supplier()
            }
        } catch (e: InterruptedException) {
            val msg = String.format("Error locking with key [%s]", key)
            logger.error(msg, e)
            Thread.currentThread().interrupt()
            throw LockInterruptException(msg, e)
        } finally {
            logger.info("Unlocking {}...", key)
            lockInstance.unlock()
            logger.info("{} unlocked.", key)
        }
    }
}